src/ring_theory/graded_algebra/basic.lean

```diff
@@ -77,4 +81,98 @@ def graded_algebra.decompose : A ≃ₐ[R] ⨁ i, 𝒜 i := alg_equiv.symm
   (graded_algebra.decompose 𝒜).symm (direct_sum.of _ i x) = x :=
 direct_sum.submodule_coe_alg_hom_of 𝒜 _ _
 
+/-- The projection maps of graded algebra-/
+def graded_algebra.proj (𝒜 : ι → submodule R A) [graded_algebra 𝒜] (i : ι) : A →ₗ[R] A :=
+(𝒜 i).subtype.comp $
+  (dfinsupp.lapply i).comp $
+  (graded_algebra.decompose 𝒜).to_alg_hom.to_linear_map
+
+@[simp] lemma graded_algebra.proj_apply (i : ι) (r : A) :
+  graded_algebra.proj 𝒜 i r = (graded_algebra.decompose 𝒜 r : ⨁ i, 𝒜 i) i := rfl
+
+/-- The support of `r` is the `finset` where `proj R A i r ≠ 0 ↔ i ∈ r.support`-/
+def graded_algebra.support [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)]
+  (r : A) : finset ι :=
+(graded_algebra.decompose 𝒜 r).support
+
+lemma graded_algebra.proj_recompose (a : ⨁ i, 𝒜 i) (i : ι) :
+  graded_algebra.proj 𝒜 i ((graded_algebra.decompose 𝒜).symm a) =
+  (graded_algebra.decompose 𝒜).symm (direct_sum.of _ i (a i)) :=
+by rw [graded_algebra.proj_apply, graded_algebra.decompose_symm_of, alg_equiv.apply_symm_apply]
+
+variable [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)]
+
+lemma graded_algebra.mem_support_iff
+  (r : A) (i : ι) :
+i ∈ graded_algebra.support 𝒜 r ↔ (graded_algebra.proj 𝒜 i r ≠ 0) :=
+begin
+  rw [graded_algebra.support, dfinsupp.mem_support_iff, graded_algebra.proj_apply],
+  simp only [ne.def, submodule.coe_eq_zero],
+end
+
+lemma graded_algebra.sum_support_decompose (r : A) :
+  ∑ i in graded_algebra.support 𝒜 r, (graded_algebra.decompose 𝒜 r i : A) = r :=
+begin
+  conv_rhs {
+    rw [←(graded_algebra.decompose 𝒜).symm_apply_apply r,
+        ←direct_sum.sum_support_of _ (graded_algebra.decompose 𝒜 r)] },
+  rw [alg_equiv.map_sum, graded_algebra.support],
+  simp_rw graded_algebra.decompose_symm_of,
+end
+
+lemma graded_algebra.mul_proj (r r' : A) (i : ι) :
+  (graded_algebra.decompose 𝒜 (r * r') i : A) =
+  ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i)
+    ((graded_algebra.support 𝒜 r).product (graded_algebra.support 𝒜 r')),
+    (graded_algebra.decompose 𝒜 r ij.1 : A) * (graded_algebra.decompose 𝒜 r' ij.2 : A) :=
+begin
```

Comment:
I think we're missing the helper lemma that:
```lean
lemma direct_sum.mul_eq_sum_support_ghas_mul {ι : Type*} {M : ι → Type*} [add_monoid ι] [decidable_eq ι]
  [Π i, add_comm_monoid (M i)] [direct_sum.gsemiring M]
  [Π (i : ι) (x : M i), decidable (x ≠ 0)]
  (a a' : ⨁ i, M i) (i : ι) :
  a * a' =
    ∑ (ij : ι × ι) in (dfinsupp.support a).product (dfinsupp.support a'),
      direct_sum.of _ _ (graded_monoid.ghas_mul.mul (a ij.fst) (a' ij.snd)) :=
begin
  change direct_sum.mul_hom _ a a' = _,
  dsimp [direct_sum.mul_hom, direct_sum.to_add_monoid, dfinsupp.lift_add_hom_apply],
  simp only [dfinsupp.sum_add_hom_apply, dfinsupp.sum, dfinsupp.finset_sum_apply,
    add_monoid_hom.coe_sum, finset.sum_apply, add_monoid_hom.flip_apply,
    add_monoid_hom.comp_hom_apply_apply, add_monoid_hom.comp_apply,
    direct_sum.gmul_hom_apply_apply],
  rw finset.sum_product,
end
```
which belongs in `direct_sum/ring`.

Reply:
With that you can make most of the progress on this lemma with
```
  dsimp only [graded_algebra.proj_apply, graded_algebra.support],
  generalize : (graded_algebra.decompose 𝒜) r = a,
  generalize : (graded_algebra.decompose 𝒜) r' = a',
  rw [alg_equiv.map_mul, direct_sum.mul_eq_sum_support_ghas_mul],
  rw [dfinsupp.finset_sum_apply, submodule.coe_sum],
```

Reply:
I think the missing piece ends up being that on a `set_like.graded_monoid` ,`(direct_sum.of _ i x j : A) = if i = j then (x : A) else 0`

Reply:
In `direct_sum.mul_eq_sum_support_ghas_mul`, the compiler is complaining
> invalid simplification lemma 'finset.sum_apply'.

Reply:
Weird

Reply:
I tried deleting `finset.sum_apply`, but I think the simplifier stucked at `dfinsupp.finset_sum_apply`

Reply:
Ah, you're missing an import of `algebra.big_operators.pi`.